contracts/token/ERC721/extensions/ERC721Wrapper.sol

```diff
@@ -0,0 +1,101 @@
+ * wrapping of an existing "basic" ERC721 into a governance token.
+ *
+ * _Available since v4.9.0_
+ */
+abstract contract ERC721Wrapper is ERC721, ERC721Holder {
+    IERC721 private immutable _underlying;
+
+    // Kept as bytes12 so it can be packed with an address
+    // Equal to 0xb125e89df18e2ceac5fd2fa8
+    bytes12 public constant WRAPPER_ACCEPT_MAGIC = bytes12(keccak256("WRAPPER_ACCEPT_MAGIC"));
+
+    constructor(IERC721 underlyingToken) {
+        _underlying = underlyingToken;
+    }
+
+    /**
+     * @dev Allow a user to deposit underlying tokens and mint the corresponding tokenIds.
+     */
+    function depositFor(address account, uint256[] memory tokenIds) public virtual returns (bool) {
+        bytes memory data = abi.encodePacked(WRAPPER_ACCEPT_MAGIC, account);
+
+        uint256 length = tokenIds.length;
+        for (uint256 i = 0; i < length; ++i) {
+            underlying().safeTransferFrom(_msgSender(), address(this), tokenIds[i], data);
+        }
+
+        return true;
+    }
+
+    /**
+     * @dev Allow a user to burn wrapped tokens and withdraw the corresponding tokenIds of the underlying tokens.
+     */
+    function withdrawTo(address account, uint256[] memory tokenIds) public virtual returns (bool) {
+        uint256 length = tokenIds.length;
+        for (uint256 i = 0; i < length; ++i) {
+            uint256 tokenId = tokenIds[i];
+            require(_isApprovedOrOwner(_msgSender(), tokenId), "ERC721Wrapper: caller is not token owner or approved");
+            _burn(tokenId);
+            // Checks were already performed at this point, and there's no way to retake ownership or approval from
+            // the wrapped tokenId after this point, so it's safe to remove the reentrancy check for the next line.
+            // slither-disable-next-line reentrancy-no-eth
+            underlying().safeTransferFrom(address(this), account, tokenId);
+        }
+
+        return true;
+    }
+
+    /**
+     * @dev Overrides {IERC721Receiver-onERC721Received} to allow minting on direct ERC721 transfers to
+     * this contract.
+     *
+     * In case there's data attached, it validates that the sender is aware of this contract's existence and behavior
+     * by checking a magic value (`WRAPPER_ACCEPT_MAGIC`) in the first 12 bytes. If it also matches, the rest 20
+     * bytes are used as an address to send the tokens to.
+     *
+     * WARNING: Doesn't work with unsafe transfers (eg. {IERC721-transferFrom}). Use {ERC721Wrapper-_recover}
+     * for recovering in that scenario.
+     */
+    function onERC721Received(
+        address,
+        address from,
+        uint256 tokenId,
+        bytes memory data
+    ) public override returns (bytes4) {
+        require(address(underlying()) == _msgSender(), "ERC721Wrapper: caller is not underlying");
+        if (data.length == 32 && WRAPPER_ACCEPT_MAGIC == bytes12(data)) {
+            from = address(bytes20(bytes32(data) << 96));
+        }
```

Comment:
In the current implementation, if the first 12 bytes are not the right ones we disregard it completely, and mint to the original from. I'd revert.

```suggestion
        if (data.length == 32) {
            (bytes12 magic, from) = abi.decode(data, (bytes12, address));
            require(magic == WRAPPER_ACCEPT_MAGIC, "ERC721Wrapper: invalid data format");
        }
```
We may also want to revert if the data.length is > 32

Reply:
I would even add 
```solidity
} else {
    require(data.length == 0, "ERC721Wrapper: invalid data format");
}

Reply:
The case in which the data is malformed is still safe because it mints to the original owner, so I wouldn't worry too much (he can always unwrap), and we allow noops, any reasons for adding this?

Reply:
If the sender is a smart contract, that thinks it unwrapped to someone else, and remove the token for its internal tracking, that might be bad.

As a general rule I think that if we ever get extra instruction that we can't process we should refuse to do anything, letting the caller know that the instructions are not clear.

Reply:
@frangio ?

Reply:
For context, the reason we added the magic value in the first place was because we didn't think we could make any assumption about the data that a sender puts in the data field. I proposed a possible scenario where a wallet uses the data field to track transfers done with that wallet software (independently of whether that's a good idea). In that case we wouldn't want to accidentally transfer to a random location because the sender happened to put data in there.

In that scenario, where the data is used for tracking purposes, my thinking was we would want to ignore the data if it doesn't contain the magic value.

However, I find this argument more convincing:
>if we ever get extra instruction that we can't process we should refuse to do anything, letting the caller know that the instructions are not clear.

So I would accept @Amxx's suggestion.

Reply:
I'm moving forward with the suggestion but I just wanted to share a meme and see if everybody gets happier plz

![325669940_2200356000148602_3945984664946041884_n](https://user-images.githubusercontent.com/33379285/217669661-459c03a7-51b8-4489-9eb5-0c5441c13e6d.jpg)